### DISCUSSION TITLE
Client Authentication Failed

### DISCUSSION THREAD

**Discussion by simonwjy** [ory/hydra#2671] [Q&A] [answered]
Hi Expert,

Our project setup an hydra server and it works fine with our hydra client. Recently we added another hydra client and then we got an issue when we try to get the token from token endpoint. Following is out hydra client configuration:

grant_types: ["client_credentials","authorization_code","refresh_token"]
response_types: ["id_token","code"]
scope: "openid offline offline_access",
token_endpoint_auth_method: "client_secret_post"

We use the same configuration for our first hydra client and it works fine. But when we try to get access token for the second hydra client we got the following error

**"The OAuth 2.0 Client supports client authentication method "client_secret_post", but method "client_secret_basic" was requested. You must configure the OAuth 2.0 client's "token_endpoint_auth_method" value to accept "client_secret_basic"**

We tried to change it to "client_secret_basic" for token_endpoint_auth_method but then it comes with another error:
**"The OAuth 2.0 Client supports client authentication method "client_secret_basic", but method "client_secret_post" was requested"**

I also try to specify AuthStyle as AuthStyleInParams but it still get the same error. Seems somewhere using two authentication ways to communicate with hydra but i have no idea where does it happen

Please let me know if i need to provide more information or you have any idea to resolve this issue.

Many thx !!!

Best Regards,
Simon
[dev-hydra.log](https://github.com/ory/hydra/files/6970322/dev-hydra.log)

**Comment by vinckr** [comment, marked as answer]
Can you check that you are using the right client and method?
`hydra clients get --endpoint http://ory-hydra <the-client-id>`
The result shows you the whole client (excluding its secret).

**Reply by simonwjy** [reply]
Hi Vinckr,

Yes we are using the right client and method, following is the output. Besides when we change the token auth endpoint method the error message change as well which i posted above so i think the client should be fine?

The thing confusing me is we have another hydra client using same configuration (token_endpoint_auth_method is client_secret_post as well) and we can get access token for that hydra client. Is there any other reason can cause this issue? why hydra is asking  client_secret_basic method when our client using client_secret_post and then after we change to client_secret_basic and then it says client_secret_post was requested 😂. 

BTW the version we are using for hydra is v1.8.5 and the hydra-client-go v1.9.2. Is it possible because the version?

Thank you.

`{
    "client_id": "xlo-dev-1-client-dp",
    "client_name": "",
    "redirect_uris": [
        "https://xilo-distributor-web.dev-1.xlo.build.dlt/api/v1/user/callback"
    ],
    "grant_types": [
        "client_credentials",
        "authorization_code",
        "refresh_token"
    ],
    "response_types": [
        "id_token",
        "code"
    ],
    "scope": "openid offline offline_access",
    "audience": [],
    "owner": "xlo-dev-1-client-dp/services-xlo-dev-1",
    "policy_uri": "",
    "allowed_cors_origins": [],
    "tos_uri": "",
    "client_uri": "",
    "logo_uri": "",
    "contacts": [],
    "client_secret_expires_at": 0,
    "subject_type": "public",
    "jwks": {},
    "token_endpoint_auth_method": "client_secret_post",
    "userinfo_signed_response_alg": "none",
    "created_at": "2021-08-11T12:35:54.128324Z",
    "updated_at": "2021-08-11T12:35:54Z",
    "metadata": {}
}`

**Reply by vinckr** [reply]
Hey, 
I just tried to reproduce with a quickstart docker-compose and your client configuration and it worked fine with `client_secret_basic`.
I was using `v1.10.2`. Maybe it is related to the older version, 1.9 was quite a big update. But then again I couldn't find anything directly related in a quick skim. 
Is doing a test of your setup on the latest version an option?
Are you doing a client credentials grant or auth code flow?

**Reply by aeneasr** [reply]
Hello, the problem is probably that your credentials are wrong. Some clients try to use POST first, then HTTP Basic Authorization, if the credentials are incorrect.

Please ensure the credentials are correct. Best idea is to recreate the client in Ory Hydra and try again.

**Reply by simonwjy** [reply]
Hi @aeneasr , i checked the credential it's correct and i used it to call /oauth2/token api by using client_credential grant type and i can get the access token.

I just observe one thing that because both hydra clients are using same hydra admin url and the public url are in different domains. when i trigger the accept login request i found that the redirect url is using the public url from another hydra client. Is this could be the reason to cause this issue? i will keep investigate this and please let me know if you have any idea on this, thank you!

**Reply by simonwjy** [reply]
Hi @vinckr, thank you for your time to reproduce it in your local. It works fine in my local but not sure why it's not working in our dev environment. If there is no other solution we might think about to upgrade the hydra version.

We are following the auth code flow. I found that when i trigger the accept login request the redirect url is refer to the public url from another hydra client not the one i'm requesting now. Do you have any idea about this and please let me know if you need more information from my side.

thank you!!

**Reply by simonwjy** [reply]
@vinckr  @aeneasr  just wondering can different hydra clients share same hydra admin url? will this cause any issue when it handle redirect 🤔 thank you

**Reply by vinckr** [reply]
Hey @SimonWangJY,
apologies for the late answer :D 
I am going through older discussions, is this still an issue for you?
